kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -355,236 +354,197 @@ uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId() const {
-                                           eg_state->event_group_id_batch.end());
-    }
-    // find and update the offsets
-    // We iterate over event_group_id_batch in the reverse order, and update the private/public offset if the event
-    // group id in event_group_id_batch is found in private_event_group_ids/public_event_group_ids respectively.
-    // In the worst case, we need to read the entire event_group_id_batch (of size kBatchSize) to find the offset
-    bool has_pub_offset_updated = false;
-    bool has_pvt_offset_updated = false;
-    for (auto r_it = eg_state->event_group_id_batch.rbegin(); r_it != eg_state->event_group_id_batch.rend(); r_it++) {
-      if (has_pub_offset_updated && has_pvt_offset_updated) break;
-      if (!has_pub_offset_updated) {
-        if (auto pub_it = std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it);
-            pub_it != public_event_group_ids.end()) {
-          eg_state->public_offset += ++pub_it - public_event_group_ids.begin();
-          has_pub_offset_updated = true;
-          LOG_DEBUG(logger_,
-                    "Updated public_offset: " << eg_state->public_offset
-                                              << " public_event_group_ids size: " << public_event_group_ids.size());
-        }
-      }
-      if (!has_pvt_offset_updated) {
-        if (auto pvt_it = std::find(private_event_group_ids.begin(), private_event_group_ids.end(), *r_it);
-            pvt_it != private_event_group_ids.end()) {
-          eg_state->private_offset += ++pvt_it - private_event_group_ids.begin();
-          has_pvt_offset_updated = true;
-          LOG_DEBUG(logger_,
-                    "Updated private_offset: " << eg_state->private_offset
-                                               << " private_event_group_ids size: " << private_event_group_ids.size());
-        }
-      }
-    }
+  if (current_ext_eg_id == external_event_group_id) {
+    return {current_global_eg_id, true, current_pvt_eg_id, external_event_group_id - current_pvt_eg_id};
   }
-  // let's not return an invalid event group id
-  if (eg_state->event_group_id_batch.empty()) return std::nullopt;
 
-  // Reset the iterator so that we read from event_group_id_batch from the beginning
-  eg_state->it = eg_state->event_group_id_batch.begin();
+  // At this point, we exhausted all private entries => it has to be a public event group
+  uint64_t global_eg_id;
 
-  ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
-  LOG_DEBUG(logger_, "Updated event group id list batch size: " << eg_state->event_group_id_batch.size());
-  return *eg_state->it++;
+  if (external_event_group_id < current_ext_eg_id) {
+    if (current_pvt_eg_id == private_start) {
+      std::tie(global_eg_id, std::ignore) = getValueFromTagTable(
+          kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(external_event_group_id));
```

Comment:
I think it would be better if we have a method/function that creates the tag such that we don't need to manually concatenate strings. Can be done in a separate commit or PR, just an idea.

Reply:
Good idea. Given that the concatenation is only used when we query the tag-table, we can overload `getValueFromTagTable` with `getValueFromTagTable(std::string tag, uint64_t event_group_id)`.
I will add this as a follow-up todo.